### PR TITLE
JAMES-2586 Implement PostgresThreadIdGuessingAlgorithm

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraThreadIdGuessingAlgorithmTest.java [hunk 1]
```diff
@@ -94,8 +94,10 @@ protected MessageId initOtherBasedMessageId() {
     }
 
     @Override
-    protected Flux<Void> saveThreadData(Username username, Set<MimeMessageId> mimeMessageIds, MessageId messageId, ThreadId threadId, Optional<Subject> baseSubject) {
-        return threadDAO.insertSome(username, hashMimeMessagesIds(mimeMessageIds), messageId, threadId, hashSubject(baseSubject));
+    protected void saveThreadData(Username username, Set<MimeMessageId> mimeMessageIds, MessageId messageId, ThreadId threadId, Optional<Subject> baseSubject) {
+        threadDAO.insertSome(username, hashMimeMessagesIds(mimeMessageIds), messageId, threadId, hashSubject(baseSubject))
+            .then()
+            .block();
     }
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/CassandraThreadIdGuessingAlgorithmTest.java:101]: Why?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresMailboxFactory.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    @Inject
+    @Singleton
+    public PostgresMailboxFactory(PostgresExecutor.Factory executorFactory) {
+        this.executorFactory = executorFactory;
+    }
+
+    public PostgresThreadDAO createThreadDAO(Optional<Domain> domain) {
+        return new PostgresThreadDAO(executorFactory.create(domain));
+    }
+}
```
Comment (chibenwa) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresMailboxFactory.java:43]: It is not a `MailboxFactory` if it builds ThreadDAO...

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithm.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            .andCriteria(SearchQuery.threadId(threadId))
+            .sorts(new SearchQuery.Sort(SearchQuery.Sort.SortClause.Arrival, SearchQuery.Sort.Order.NATURAL))
+            .build();
+
+        MultimailboxesSearchQuery expression = MultimailboxesSearchQuery
+            .from(searchQuery)
+            .build();
+
+        return Flux.from(mailboxManager.search(expression, session, Integer.MAX_VALUE))
+            .switchIfEmpty(Mono.error(() -> new ThreadNotFoundException(threadId)));
```
Comment (chibenwa) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithm.java:93]: I would love to be able to do this against PG only.
Here I understand this would be done against OpenSearch.

I bet this would be as easy as reading the `thread` table by `username` and `threadId` ?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+public class PostgresThreadDAO {
+    private final PostgresExecutor postgresExecutor;
+
+    public PostgresThreadDAO(PostgresExecutor postgresExecutor) {
+        this.postgresExecutor = postgresExecutor;
+    }
+
+    public Mono<Void> insertSome(Username username, Set<Integer> hashMimeMessageIds, MessageId messageId, ThreadId threadId, Optional<Integer> hashBaseSubject) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.batch(
+            hashMimeMessageIds.stream().map(hashMimeMessageId -> dslContext.insertInto(TABLE_NAME)
```
Comment (chibenwa) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:54]: Use multi insert instead: https://stackoverflow.com/questions/20815028/how-do-i-insert-multiple-values-into-a-postgres-table-at-once

CF https://www.jooq.org/doc/latest/manual/sql-building/sql-statements/insert-statement/insert-values/

```
create.insertInto(AUTHOR,
        AUTHOR.ID, AUTHOR.FIRST_NAME, AUTHOR.LAST_NAME)
      .values(100, "Hermann", "Hesse")
      .values(101, "Alfred", "Döblin")
      .execute();
```

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                .set(THREAD_ID, ((PostgresMessageId) threadId.getBaseMessageId()).asUuid())
+                .set(HASH_BASE_SUBJECT, hashBaseSubject.orElse(null)))
+                .collect(ImmutableList.toImmutableList()))));
+    }
+
+    public Flux<Pair<Optional<Integer>, ThreadId>> findThreads(Username username, Set<Integer> hashMimeMessageIds) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select(THREAD_ID, HASH_BASE_SUBJECT)
+            .from(TABLE_NAME)
+            .where(USERNAME.eq(username.asString()))
+            .and(HASH_MIME_MESSAGE_ID.in(hashMimeMessageIds))))
```
Comment (chibenwa) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:67]: Can I break this by supplying a message with an absurdly large number of references?

Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:67]: How many?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithm.java (hunk 3 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithm.java:93]: In the current implementation, data will be sorted by arrival date. In postgres implementation, sorting data by MESSAGE_ID could be good option. And to do that, MESSAGE_ID must be generated by UuidCreator.getTimeOrderedEpoch(). WDYT?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresMailboxFactory.java (hunk 2 above)
Comment (Arsnael) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresMailboxFactory.java:43]: Yeah I don't understand that class either... Why you don't follow the current pattern? Should have a Factory subclass in PostgresThreadDAO, to make `PostgresThreadDAO.Factory`, like the other generated factories in `PostgresMailboxSessionMappoerFactory`?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                .column(MESSAGE_ID)
+                .column(THREAD_ID)
+                .column(HASH_BASE_SUBJECT)
+                .constraint(DSL.primaryKey(USERNAME, HASH_MIME_MESSAGE_ID, MESSAGE_ID))))
+            .supportsRowLevelSecurity()
+            .build();
+
+        PostgresIndex MESSAGE_ID_INDEX = PostgresIndex.name("thread_message_id_index")
+            .createIndexStep((dsl, indexName) -> dsl.createIndexIfNotExists(indexName)
+                .on(TABLE_NAME, MESSAGE_ID));
```
Comment (Arsnael) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java:59]: Wouldn't it make sense to have THREAD_ID being indexed as well?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+        PostgresTable TABLE = PostgresTable.name(TABLE_NAME.getName())
+            .createTableStep(((dsl, tableName) -> dsl.createTableIfNotExists(tableName)
+                .column(USERNAME)
+                .column(HASH_MIME_MESSAGE_ID)
+                .column(MESSAGE_ID)
+                .column(THREAD_ID)
+                .column(HASH_BASE_SUBJECT)
+                .constraint(DSL.primaryKey(USERNAME, HASH_MIME_MESSAGE_ID, MESSAGE_ID))))
+            .supportsRowLevelSecurity()
```
Comment (quantranhong1999) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java:54]: No need RLS IMO, as information in this table is hashed and not critical data subject to tenant data leak issue.

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java (hunk 6 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java:59]: Will be added if https://github.com/apache/james-project/pull/1941#discussion_r1464356005 is implemented

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresMailboxFactory.java (hunk 2 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresMailboxFactory.java:43]: I think it would be better to gather all DAOs creation in one feactory instead of one factory per DAO. WDYT?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java (hunk 7 above)
Comment (quantranhong1999) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java:54]: Yeah. I am not fully aware of the RLS scope either...

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresMailboxFactory.java (hunk 2 above)
Comment (Arsnael) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresMailboxFactory.java:43]: I think the factories being subclasses of their DAO is easier to follow IMO

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java (hunk 5 above)
Comment (chibenwa) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:67]: Maybe we can bathc them by group of 16?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java (hunk 4 above)
Comment (chibenwa) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:54]: Other opinions?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java (hunk 7 above)
Comment (chibenwa) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java:54]: Linked to a user -> RLS is doable ?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java (hunk 5 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:67]: It seems that Postgres is still able to handle hundreds elements in IN clause without any issues. What is maximum length of hashMimeMessageIds?

Comment (quantranhong1999) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:67]: > What is maximum length of hashMimeMessageIds?

Can not say for sure. If the conversation keeps being looped then the length could be very big :D

On IN limitation, I remember @vttranlina apply it somewhere in the mailbox component?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java (hunk 4 above)
Comment (quantranhong1999) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:54]: IMO we should use a single INSERT Statement with Multiple Sets of Values.

ChatGPT:
```
1. Single INSERT Statement with Multiple Sets of Values:

sql
Copy code
INSERT INTO your_table (column1, column2, column3)
VALUES
    (value1_1, value1_2, value1_3),
    (value2_1, value2_2, value2_3),
    ...
    (valueN_1, valueN_2, valueN_3);
Pros:

Single statement execution, which can be more efficient and have lower overhead.
Easier to read and maintain for a small number of rows.

Generally performs better than a batch of individual INSERT statements.

Atomic execution, meaning either all rows are inserted, or none are.

Better for inserting large sets of data.

Cons:

Limited by the maximum allowed query size, which might lead to issues if the number of rows is very large.
Might become harder to read and maintain as the number of rows increases.

2. Batch INSERT Statements:

sql
Copy code
INSERT INTO your_table (column1, column2, column3) VALUES (value1_1, value1_2, value1_3);
INSERT INTO your_table (column1, column2, column3) VALUES (value2_1, value2_2, value2_3);
...
INSERT INTO your_table (column1, column2, column3) VALUES (valueN_1, valueN_2, valueN_3);
Pros:

Easier to generate dynamically if the data is coming from various sources or systems.
More control over the execution of each individual statement.
Cons:

Higher overhead due to multiple statement executions.
Might be slower for a large number of rows compared to a single INSERT statement.
Not as readable or maintainable for a large number of rows.
Each statement is executed independently, meaning some rows may be inserted while others fail.
```

TLDR:
- multi values INSERT seems to have better performance than a BATCH statement.
- multi values INSERT is atomic, while BATCH statement is not.

Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:54]: I asked ChatGPT by Vietnamese and it gave the opposite answer :((

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java (hunk 6 above)
Comment (chibenwa) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java:59]: Should be added given that https://github.com/apache/james-project/pull/1941#discussion_r1464356005 will be implemented.

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java (hunk 4 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:54]: one more thing, batch should be used together with prepared statement, not alone

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithm.java (hunk 3 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithm.java]: updated

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java (hunk 5 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java]: partition added

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java (hunk 6 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadModule.java]: added

**Review by Arsnael — APPROVED** [review]

File: mailbox/postgres/src/test/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithmTest.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    @Override
+    protected void saveThreadData(Username username, Set<MimeMessageId> mimeMessageIds, MessageId messageId, ThreadId threadId, Optional<Subject> baseSubject) {
+        PostgresThreadDAO threadDAO = threadDAOFactory.create(username.getDomainPart());
+        threadDAO.insertSome(username, hashMimeMessagesIds(mimeMessageIds), PostgresMessageId.class.cast(messageId), threadId, hashSubject(baseSubject)).block();
+    }
+
+    @Test
+    void givenAMailInAThreadThenGetThreadShouldReturnAListWithOnlyOneMessageIdInThatThread() throws MailboxException {
```
Comment (chibenwa) [inline mailbox/postgres/src/test/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithmTest.java:101]: Why those tests are not part of the contract?

**Review by vttranlina — APPROVED** [review]

File: mailbox/postgres/src/test/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithmTest.java (hunk 8 above)
Comment (hungphan227) [reply mailbox/postgres/src/test/java/org/apache/james/mailbox/postgres/PostgresThreadIdGuessingAlgorithmTest.java:101]: Instead of using threadDAO, getMessageIdsInThread method in CassandraThreadIdGuessingAlgorithm use mailboxManager to get data. The contract use messageMapper to create data for testing getMessageIdsInThread method. 
PostgresThreadIdGuessingAlgorithm use threadDAO to get data in getMessageIdsInThread method. Therefore, I also use threadDAO to create data for testing getMessageIdsInThread method. Thus, I have to override givenAMailInAThreadThenGetThreadShouldReturnAListWithOnlyOneMessageIdInThatThread of the contract

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java (hunk 4 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:76]: ChatGPT:

1.  **Single Insert Query with Multiple Rows:**
    
    *   **Advantages:**
        
        *   Simplicity: Writing a single insert query for multiple rows is straightforward and easy to implement.
        *   Fewer round-trips: You send one query to the database server, reducing the number of round-trips between your application and the database.
    *   **Considerations:**
        
        *   Scalability: For a large number of rows, a single insert query can become unwieldy and may lead to performance issues.
        *   Limited reuse: Each row's values need to be explicitly specified in the query, limiting the reuse of the query for different sets of data.
        
2.  **Batch Insert with Prepared Statements:**
    
    *   **Advantages:**
        
        *   Performance: Prepared statements can be more efficient, especially when inserting a large volume of data, as the query is precompiled and can be reused with different parameter values.
        *   Parameterized queries: Prepared statements allow you to use parameterized queries, enhancing security by protecting against SQL injection.
    *   **Considerations:**
        
        *   Complexity: Implementing batch inserts with prepared statements may require more code and complexity in your application.
        *   Overhead for small datasets: For a small number of rows, the overhead of using prepared statements might outweigh the performance benefits.

**Recommendations:**

*   For small to moderate datasets, a single insert query with multiple rows may be simpler and sufficient.
*   For large datasets or when performance is critical, consider using batch inserts with prepared statements. This can be particularly beneficial when inserting thousands or millions of rows, as it reduces the overhead of repeatedly sending similar queries.

Always profile and test the performance of your specific use case to determine the most suitable approach. Additionally, keep in mind that other factors, such as indexing, database configuration, and network latency, can also influence the overall performance.

Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresThreadDAO.java:76]: In conclusion, Single Insert Query with Multiple Rows may not work well with a large number of rows. Single Insert Query with Multiple Rows ensure atomicity while Batch Insert with Prepared Statements does not.

Regarding atomicity, CassandraThreadIdGuessingAlgorithm does not ensure atomicity so I think it is not necessary to ensure atomicity in PostgresThreadIdGuessingAlgorithm

**Review by quantranhong1999 — APPROVED** [review]
